src/firebolt_db/firebolt_async_dialect.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+from types import ModuleType
+from typing import Any, Iterator, List, Optional, Tuple
+
+import firebolt.async_db as async_dbapi
+from firebolt.async_db import Connection
+from sqlalchemy.engine import AdaptedConnection  # type: ignore[attr-defined]
+from sqlalchemy.util.concurrency import await_only
+
+from firebolt_db.firebolt_dialect import FireboltDialect
+
+
+class AsyncCursorWrapper:
+    __slots__ = (
+        "_adapt_connection",
+        "_connection",
+        "description",
+        "await_",
+        "_cursor",
+        "_rows",
+        "arraysize",
+        "rowcount",
+    )
+
+    server_side = False
+
+    def __init__(self, adapt_connection: AsyncConnectionWrapper):
+        self._adapt_connection = adapt_connection
+        self._connection = adapt_connection._connection
+        self.await_ = adapt_connection.await_
+        self.arraysize = 1
+        self.rowcount = -1
+        self.description = None
+        self._rows: List[Tuple] = []
+
+    def close(self) -> None:
+        self._rows[:] = []
+
+    def execute(self, operation: str, parameters: Optional[Tuple] = None) -> None:
+        _cursor = self._connection.cursor()
+        self.await_(_cursor.execute(operation, parameters))
+        if _cursor.description:
+            self.description = _cursor.description
+            self.rowcount = -1
```

Comment:
Shouldn't it be vice versa? When description is None rowcount is -1, else rowcount is taken from cursor

Reply:
You're right, I assumed sqllite lib knows something I don't and copied it. Fixed it and added tests.